### PR TITLE
feat(tekton provider): Update tekton provider

### PR DESCRIPTION
**Pull request by briangleeson**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### Changes

Update the tekton pipeline provider with latest go sdk:
- Bump `github.com/IBM/continuous-delivery-go-sdk` to `v0.0.8`
- Refactor naming scheme from `tekton_pipeline_*` to `cd_tekton_pipeline_*`
- Add some new diff suppress functions
- Add `ForceNew` to a few schemas where necessary for immutable values
- Add enum handling to Trigger Property
- Add missing description for some id properties
- Added `href` property to pipeline run resource
- Update all strings after internal review
- Change `created` property to `created_at`
- **Breaking**: Change `concurrency` object property to `max_concurrent_runs` integer property
- Add support for the `preview` worker option in v2 APIs, when supported in that region. Falls back to `public` if not supported
- **Breaking**: Reshape scm source object to include service_instance_id 
- Change patch requests to use JSON merge-patch content-type
- Rename `html_url` to `runs_url` and `run_url`, for pipeline and pipelineRun resources respectively
- **Breaking**: Reshape property types to use snake case: `text`, `single_select`, `integration`, `secure`, `appconfig`
- **Breaking**: Reshape secret types to use snake case: `token_matches`, `digest_matches`, `internal_validation`
- Add support for `enable_slack_notifications` and `enable_partial_cloning` pipeline settings
- **Breaking**: Reshape enum value to remove the `default` option in favour of `value` instead
- **Breaking**: Refactored Trigger into separate request and response shapes

### Output from acceptance testing:

```
$ make testacc TEST=./ibm/service/cdtoolchain TESTARGS='-run=TestAccIBMCdToolchain*'

=== RUN   TestAccIBMCdToolchainDataSourceBasic
--- PASS: TestAccIBMCdToolchainDataSourceBasic (52.43s)
=== RUN   TestAccIBMCdToolchainDataSourceAllArgs
--- PASS: TestAccIBMCdToolchainDataSourceAllArgs (46.79s)
=== RUN   TestAccIBMCdToolchainToolAppconfigDataSourceBasic
--- PASS: TestAccIBMCdToolchainToolAppconfigDataSourceBasic (49.96s)
=== RUN   TestAccIBMCdToolchainToolAppconfigDataSourceAllArgs
--- PASS: TestAccIBMCdToolchainToolAppconfigDataSourceAllArgs (48.36s)
=== RUN   TestAccIBMCdToolchainToolArtifactoryDataSourceBasic
--- PASS: TestAccIBMCdToolchainToolArtifactoryDataSourceBasic (75.43s)
=== RUN   TestAccIBMCdToolchainToolArtifactoryDataSourceAllArgs
--- PASS: TestAccIBMCdToolchainToolArtifactoryDataSourceAllArgs (56.61s)
=== RUN   TestAccIBMCdToolchainBasic
--- PASS: TestAccIBMCdToolchainBasic (85.75s)
=== RUN   TestAccIBMCdToolchainAllArgs
--- PASS: TestAccIBMCdToolchainAllArgs (71.86s)
=== RUN   TestAccIBMCdToolchainToolPipelineBasic
--- PASS: TestAccIBMCdToolchainToolPipelineBasic (48.86s)
=== RUN   TestAccIBMCdToolchainToolPipelineAllArgs
--- PASS: TestAccIBMCdToolchainToolPipelineAllArgs (91.56s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cdtoolchain     633.398s

$ make testacc TEST=./ibm/service/cdtektonpipeline TESTARGS='-run=TestAccIBMCdTekton*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/cdtektonpipeline -v -run=TestAccIBMCdTekton* -timeout 700m
=== RUN   TestAccIBMCdTektonPipelineBasic
--- PASS: TestAccIBMCdTektonPipelineBasic (99.81s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cdtektonpipeline        105.269s
```

#### Remaining work
- [X] Update the GO sdk version used for tekton pipeline once the new version has been published
- [X] Fix the detect-secrets errors in PR build
- [X] Fix cdtektonpipeline tests

### CODE REVIEW

File: ibm/provider/provider.go
Comment (hkantare) [inline ibm/provider/provider.go:360]: All these changes in the file is removing the support for other services which were added recently

Comment (hkantare) [reply ibm/provider/provider.go:360]: Take the latest code and add your changes on the top of it

File: ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go
Comment (hkantare) [inline ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:128]: Will id be a optional argument?

File: ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go
Comment (hkantare) [inline ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: same as below a computed can't have inner attributes as Required/optional

File: website/docs/d/cd_tekton_pipeline.html.markdown
Comment (hkantare) [inline website/docs/d/cd_tekton_pipeline.html.markdown:34]: Again all attibute reference can't have Required/optional ..They are just computed in all docs

File: website/docs/d/cd_tekton_pipeline.html.markdown
Comment (hkantare) [inline website/docs/d/cd_tekton_pipeline.html.markdown:19]: Datasource requires "pipeline_id" not "id" as argument

File: ibm/provider/provider.go
Comment (briangleeson) [reply ibm/provider/provider.go]: Fixed

File: ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go
Comment (briangleeson) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: service_instance_id is only marked as `required` now, not computed

File: ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go
Comment (briangleeson) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:128]: Actually no. The definition id would be computed upon creation and would not change. I'll remove the optional arg

File: ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go
Comment (briangleeson) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: I can update the description to make it clearer what this property is for

File: website/docs/d/cd_tekton_pipeline.html.markdown
Comment (briangleeson) [reply website/docs/d/cd_tekton_pipeline.html.markdown:34]: Not sure I follow, can you clarify your meaning?

Comment (briangleeson) [reply website/docs/d/cd_tekton_pipeline.html.markdown:34]: build_number no longer has Optional param in the doc

File: ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go
Comment (briangleeson) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: Thanks for the clarification. This is a mistake, definitions should not be marked computed here. It appears to be an issue with [the generator](https://github.ibm.com/CloudEngineering/openapi-sdkgen) we use to create the provider.

I will look at updating the provider to remove erroneous Computed options such as the example you posted

Comment (briangleeson) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: I think the issue here is that most of these properties are ones that *are* computed but only on the return object. They are not needed, nor should they be used, when making changes to a Tekton pipeline resource.

Would it make sense then to remove all of the return object properties and only include them in the data source for Tekton pipeline? That makes more sense to me

Comment (hkantare) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: We generally keep all computed properties in resource also else it makes us to dependent on one more datasouce if any one of properties need to referred in another resource.. 
May be we can work with generator team if a outer attribute is marked as computed then by default can we mark the inner attributes as computed irrespective
I think it will be as of readability issue for developers

Comment (briangleeson) [reply ibm/service/cdtektonpipeline/resource_ibm_cd_tekton_pipeline.go:121]: I'd like to hold off merging for the moment. The more I've worked on this and reviewed with my team, the more changes we'd like to make

**Comment by briangleeson** [comment]
Moving this back to draft for now. We have more changes identified that we'd like to add to this before moving forward

**Comment by briangleeson** [comment]
- Large amount of String updates added
- Refactored `created`, `updated` and `concurrency` properties
- Added `href` property to pipeline run resource
- Updated docs

**Comment by briangleeson** [comment]
TODO
- [ ] resolve merge conflicts
- [ ] update GO sdk version once built and available

**Comment by briangleeson** [comment]
@hkantare we are close to completing this piece of work. I'm wondering about the process to get this change (once approved) out into a provider release. Can you point me to a doc that describes this process?

**Comment by briangleeson** [comment]
TODO: update continuous-delivery-go-sdk version once the new build is available

**Comment by briangleeson** [comment]
@hkantare I've marked the PR ready for review again. Let me know if you have any feedback, thanks

**Review by hkantare — APPROVED** [review]

**Pull request merged**
